plugins/node.d.linux/psi

```diff
@@ -0,0 +1,94 @@
+#!/bin/sh
+
+: << '=cut'
+
+=head1 NAME
+
+psi - Plugin to graph Linux pressure stall information
+
+=head1 APPLICABLE SYSTEMS
+
+Linux kernels since 4.20 with CONFIG_PSI=y.
+
+=head1 INTERPRETATION
+
+When resources are contended, workloads experience latency spikes, throughput
+losses, and run the risk of OOM kills.
+
+Linux's PSI feature, introduced in kernel 4.20, identifies and quantifies the
+disruptions caused by resource contention.
+
+=head1 AUTHOR
+
+Copyright (C) 2022 Chris Down <chris@chrisdown.name>
+
+=head1 LICENSE
+
+GNU GPLv2
+
+=head1 MAGIC MARKERS
+
+  #%# family=auto
+  #%# capabilities=autoconf
+
+=cut
+
+psi_dir=/proc/pressure
+
+autoconf() {
+    if [ -d "$psi_dir" ]; then
+        echo yes
+    else
+        echo "no ($psi_dir not found)"
+    fi
+}
+
+config() {
+    echo "graph_title PSI"
+    echo "graph_args --upper-limit 100 -l 0"
+    echo "graph_vlabel %"
+    echo "graph_category system"
+    echo "graph_info This graph shows resource contention on the machine." \
+        "Values are recorded from the 'avg10' counter." \
+        "%full cpu is excluded as it's meaningless at the system level."
+
+    for res in memory cpu io; do
+        while read -r psi_type _; do
+            if [ "$res" = cpu ] && [ "$psi_type" = full ]; then
+                # See comment in get_data
+                continue
+            fi
+            printf '%s%spressure.label %%%s %s pressure\n' \
+                "$res" "$psi_type" "$psi_type" "$res"
+            printf '%s%spressure.info Percent %s time in "%s" contention\n' \
+                "$res" "$psi_type" "$res" "$psi_type"
+            printf '%s%spressure.type DERIVE\n' "$res" "$psi_type"
```

Comment:
As the author of "the other psi plugin", I've got to say, I'm pretty impressed by the code efficiency of this plugin. There's still something to learn for me.

I'm curious though, why you chose to use `DERIVE` for the "average" metrics. Maybe I got something wrong from the documentation (which is not unlikely), but as far as I know, these values are in percent. Is there a major benefit in watching just the rate of change? I can see that for the "total" metrics, which are only ever increasing and reach astronomical values that hide all details when viewed as is. But for the "averages"?

Also, is there a reason why you focus on the 10 second average? After all, by default munin collects metrics every 5 minutes (which can be changed of course). Therefore the 300 second average would seam more logical to me. However, I don't have a lot of experience with the psi values in general yet. So I'm probably missing something important here (which is why I'm asking).

Reply:
I chose `DERIVE` because, as I understand it, derive + min 0 prevents dropping below zero due to sampling shenanigans.

The code focuses on the 10 second average because while the counters are aggregated at 5 minutes, they are collected more frequently, as I understand it. As such, focusing on avg300 is likely to obscure resource contention spikes.

I am by no means a Munin expert though, so I may be wrong about the derive + min or collection interval stuff. :-)

Reply:
Well, as far as I know, derive does what it says. It derives the function. So instead of the actual value, you get the rate of change compared to the previous value(s). This is very useful for metrics, where you don't care as much about the actual value but more about it's direction.

For example the "total" metrics for the psi, which are (as far as I know) counter values. So the only ever go up. Using a derive, one can easily see when and by what rate they rise. However, once you reboot the machine, the counter gets reset and starts at zero again. For a graph with "derive" this would mean a huge negative spike, that actually has no meaning, because it's just caused by a counter reset. Therefore using `min 0` prevents this.

However when looking at the average metrics, we are looking at the percentages of an average over time. These values naturally go back to zero every now and then. Well, I guess that's an understatement and would probably be a sign of problems. Let me rephrase that: "These values are not zero every now and then". Whatever, the point is, that it's normal for these values to go back to zero and therefor would cut of that portion of the derived graph (when using `min 0`) and give a false impression. Also, since these values are in percent already, I don't see the point in using `DERIVE` at all. I personally can grasp values in percent very well. Therefore I was wondering what the benefit of deriving these metrics is. Since you are the one with way more experience about psi than I am, I figured there might be a good reason for that. So I asked.

Same goes for the choice to focus on the 10 second average. Yes the 300 second average is an average over a bigger time span and therefore looses details. On the other hand the 10 second average only shows the spikes that are present in the average in the 10 seconds before the plugin was invoked. The other 290 seconds between two plugin calls are completely ignored. I guess both have their benefits, maybe not. I'm not an expert on psi yet. For my plugin is solved this problem by including all metrics by default and giving the option to exclude ones that you don't care about by configuration. Which to be fair is a pretty lame approach, since it spared me the time for looking deeper into how psi works.

So don't get me wrong. It's not meant as bad critique. I just stumbled across these things that seamed odd to me. But since I have to admit, that I'm not an expert on the psi metrics, I figured there is probably something I'm missing and took the opportunity to learn something.

Reply:
Thank you for your exchange of thoughts!

Regarding the `avg300`: I think, this should be the proper choice here. All values will be lower, but this does not prevent us from seeing relative changes in proper proportion. The coverage of munin's default period of 300 seconds is preferable.
Could you please change this?